python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py

```diff
@@ -0,0 +1,158 @@
+
+        @asset_check(
+            asset=asset_key,
+            description="Evaluates freshness for targeted asset.",
+            name="freshness_check",
+        )
+        def the_check(context: AssetExecutionContext) -> AssetCheckResult:
+            current_time = pendulum.now()
+            current_timestamp = check.float_param(current_time.timestamp(), "current_time")
+            partitions_def: TimeWindowPartitionsDefinition = cast(
+                TimeWindowPartitionsDefinition,
+                check.inst_param(
+                    context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+                    "partitions_def",
+                    TimeWindowPartitionsDefinition,
+                ),
+            )
+            latest_cron_tick_time = check.not_none(
+                get_latest_completed_cron_tick(
+                    freshness_cron, current_time, partitions_def.timezone
+                ),
+                "expected there to be a previous tick of the provided cron",
+            )
+
+            time_window = partitions_def.get_prev_partition_window(latest_cron_tick_time)
+            if not time_window:
+                return AssetCheckResult(
+                    passed=True,
+                    description="No partitions have been completed yet.",
+                    severity=severity,
+                )
+            partition_key = partitions_def.get_partition_key_range_for_time_window(
+                time_window
+            ).start
+            latest_record = retrieve_latest_record(context.instance, asset_key, partition_key)
+            passed = latest_record is not None
+            expected_partition_key = partitions_def.get_partition_key_for_timestamp(
+                time_window.start.timestamp()
+            )
+            if passed:
+                last_updated_timestamp = check.float_param(
+                    get_last_updated_timestamp(latest_record), "last_updated_timestamp"
+                )
+                metadata = {
+                    "dagster/last_updated_timestamp": TimestampMetadataValue(
+                        last_updated_timestamp
+                    ),
+                }
+                last_updated_time = pendulum.from_timestamp(
+                    last_updated_timestamp, tz=partitions_def.timezone
+                )
+                if latest_record.asset_materialization:
+                    description = (
+                        f"Partition {expected_partition_key} arrived on time. "
+                        f"Materialized on {last_updated_time.to_formatted_date_string()} at "
+                        f"{last_updated_time.to_time_string()}."
+                    )
+                else:
+                    description = (
+                        f"Partition {expected_partition_key} arrived on time. "
+                        f"Latest data was from "
+                        f"{last_updated_time.to_formatted_date_string()} at "
```

Comment:
No strong preference, but could also use use `to_formatted_datetime_string` I believe.

Reply:
Something to be careful with here is timezone. When testing out similar code, I ended up in a confusing situation where this string was in a different timezone than all the other timestamps in the UI. IMO we should either:
- Show the timezone in this string
- Rely on the metadata values to show the datetime, because the frontend can use the same timezone for it as for all the other timestamps

Reply:
For consistency's sake I agree it's less goofy to keep it out of the description.